From 4000fd083dd67d7957dabf266f990dc3ee9bc1fd Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Fri, 29 Apr 2022 06:14:27 +0300
Subject: [PATCH] [SW-66675] gaudi3: add tpm erom routine

TPM eROM routine is added in ROM . The selection of tpm
run location (ROM/eROM) is done using BIT 6 in efuse offset
0x3A.

Change-Id: I052909290c87fd30fffe1416795c6df0f2e56abf
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index af9698b..e326f31 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -685,7 +685,14 @@
 		return;
 
 	if (!security_efuse_tpm_loc()) {
-		/* TODO: Add erom tpm routine*/
+		if (!boot_desc_init(HL_FW_BOOT_PTR_EROM)) {
+			arcboot_rom_status_update(
+					PROGRESS_ARCBOOT_EROM_BOOT_DESC_ERROR);
+			return;
+		}
+
+		/* Load TPM binary from EROM */
+		ext_rom_load_erom_image(HL_FW_EROM_TPM);
 	} else {
 		arcboot_rom_status_update(PROGRESS_ARCBOOT_ROM_TPM);
 
diff --git a/arch/gaudi3/arc/dev_common.h b/arch/gaudi3/arc/dev_common.h
index 1fea950..da7e00c 100644
--- a/arch/gaudi3/arc/dev_common.h
+++ b/arch/gaudi3/arc/dev_common.h
@@ -6,8 +6,10 @@
 #include <arcboot.h>
 
 #define GPIOS_PER_BANK			16
-#define TPM_CS_GPIO_PIN			10
+#define TPM_CS_GPIO_PIN			18
 
 void arcboot_rom_notify_slave(void);
 uint32_t arcboot_rom_get_num_crc(uint32_t size);
 void arcboot_rom_wait_for_fuse_data(void);
+int arcboot_rom_fuse_single_die_config(void);
+void ext_rom_load_erom_image(enum hl_fw_img_type_erom img_type);
diff --git a/arch/gaudi3/arc/recovery.c b/arch/gaudi3/arc/recovery.c
index ad7671c..582c5d2 100644
--- a/arch/gaudi3/arc/recovery.c
+++ b/arch/gaudi3/arc/recovery.c
@@ -71,7 +71,7 @@
 	return DMA_DEST_HBW;
 }
 
-static void ext_rom_load_erom_image(enum hl_fw_img_type_erom img_type)
+void ext_rom_load_erom_image(enum hl_fw_img_type_erom img_type)
 {
 	enum arcboot_rom_progress img_sts_base = PROGRESS_ARCBOOT_EROM_PCIE;
 	enum dma_error dma_err = DMA_SUCCESS;
diff --git a/include/arcboot.h b/include/arcboot.h
index 0984180..f731a1d 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -234,24 +234,29 @@
 	PROGRESS_ARCBOOT_EROM_RECOVERY_FINISH = 96,
 	PROGRESS_ARCBOOT_EROM_TPM = 97,
 	PROGRESS_ARCBOOT_EROM_TPM_DMA_START = 98,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH = 99,
-	PROGRESS_ARCBOOT_EROM_TPM_START = 100,
-	PROGRESS_ARCBOOT_EROM_TPM_FINISH = 101,
-	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 102,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_UP = 103,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 104,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 105,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 106,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 107,
-	PROGRESS_ARCBOOT_ROM_FIT_FORMAT_HEADER_CHECK_FAILED = 108,
-	PROGRESS_ARCBOOT_ROM_FIT_FORMAT_NO_DESCRIPTION = 109,
-	PROGRESS_ARCBOOT_ROM_FIT_FORMAT_NO_IMAGES_PARENT_NODE = 110,
-	PROGRESS_ARCBOOT_ROM_FIT_FORMAT_CORRUPTED_LOAD_TYPE = 111,
-	PROGRESS_ARCBOOT_ROM_FIT_FORMAT_UNSUPPORTED_LOAD_TYPE = 112,
-	PROGRESS_ARCBOOT_ROM_FIT_FORMAT_UNKNOWN_ERROR = 113,
-	PROGRESS_ARCBOOT_ROM_FIT_LOAD_FAILED = 114,
-	PROGRESS_ARCBOOT_ROM_FIT_LOAD_SEC_FAILED = 115,
-	PROGRESS_ARCBOOT_ROM_FIT_LOAD_UNKNOWN_ERROR = 116,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_CRC_FAILED = 99,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_UNALIGNED_DEST = 100,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_UNALIGNED_IMG_SIZE = 101,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_ERROR = 102,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH = 103,
+	PROGRESS_ARCBOOT_EROM_TPM_VERIFICATION_FAIL = 104,
+	PROGRESS_ARCBOOT_EROM_TPM_START = 105,
+	PROGRESS_ARCBOOT_EROM_TPM_FINISH = 106,
+	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 107,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_UP = 108,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 109,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 110,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 111,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 112,
+	PROGRESS_ARCBOOT_ROM_FIT_FORMAT_HEADER_CHECK_FAILED = 113,
+	PROGRESS_ARCBOOT_ROM_FIT_FORMAT_NO_DESCRIPTION = 114,
+	PROGRESS_ARCBOOT_ROM_FIT_FORMAT_NO_IMAGES_PARENT_NODE = 115,
+	PROGRESS_ARCBOOT_ROM_FIT_FORMAT_CORRUPTED_LOAD_TYPE = 116,
+	PROGRESS_ARCBOOT_ROM_FIT_FORMAT_UNSUPPORTED_LOAD_TYPE = 117,
+	PROGRESS_ARCBOOT_ROM_FIT_FORMAT_UNKNOWN_ERROR = 118,
+	PROGRESS_ARCBOOT_ROM_FIT_LOAD_FAILED = 119,
+	PROGRESS_ARCBOOT_ROM_FIT_LOAD_SEC_FAILED = 120,
+	PROGRESS_ARCBOOT_ROM_FIT_LOAD_UNKNOWN_ERROR = 121,
 };
 
 enum security_location {
